### PR TITLE
fix(commons/dom): treat a select's button and selectedcontent as inert

### PR DESCRIPTION
**Pull request by ItsMeSousuke**
The HTML spec makes the button and selectedcontent inside a select inert: they are excluded from the accessibility tree and cannot be focused. Without this, the customizable-select (`appearance: base-select`) markup reports a false-positive `button-name` violation.

#### Describe the changes

- `isInert` (`lib/commons/dom/is-inert.js`) now treats any `<button>` that is a direct child of a `<select>`, and any `<selectedcontent>` element, as inert, per the [HTML spec's `select` element definition](https://html.spec.whatwg.org/multipage/form-elements.html#the-select-element). This is unconditional — it doesn't check the select's rendered `appearance` — since inertness is an HTML behavior, not a CSS one. This change propagates to `isVisibleToScreenReaders` and `focusDisabled`/`isFocusable`, which already delegate to `isInert`, so `nested-interactive`, `color-contrast`, and any other rule relying on those utilities also benefit.
- Added `lib/rules/button-name-matches.js`, wrapping `no-explicit-name-required-matches` with an `isInert` check, and pointed `button-name.json`'s `matches` at it. This is needed because `axe.runVirtualRule` forces `excludeHidden: false` and bypasses the gather-time visibility filter that `axe.run` applies.
- Added unit tests covering the select button/`selectedcontent` cases (including Shadow DOM) in `test/commons/dom/is-inert.js`, `is-visible-to-screenreader.js`, `focus-disabled.js`, and `is-focusable.js`.
- Added `test/rule-matches/button-name-matches.js` and a `SerialVirtualNode` case in `test/integration/virtual-rules/button-name.js` to cover the `matches` function directly.
- Added an inapplicable select-button fixture to `test/integration/rules/button-name/button-name.html`.

#### Closes:

Closes issue #5315

### CODE REVIEW

**Comment by straker** [comment]
Thanks for the pr. This is a good first pass, but I think needs some more work before we can merge it in. I tested to see how `<button>` is treated inside a `<select>` element, and it seems that regardless of if the button is the first child of a select the button itself is never rendered in any browser and no AT can navigate to them. In the following example neither of the 3 buttons are visible / navigatable.

```html
<select>
  <button>foo</button>
  <button>bar</button>
  <option>Hello</option>
  <span><button>bar</button></span>
</select>
```

This to me suggests that a `<button>` should return `true` for the function `isHiddenForEveryone` when it is a descendant of a `<select>`. An interesting thing though is that when the `<button>` is a child of an `<option>` it is visible but treated as the `<option>` rather than as a separate button, so I think it's fine if we treat the button as hidden still.

I think the way we want to handle this is to add a new [visibility method](https://github.com/dequelabs/axe-core/blob/develop/lib/commons/dom/visibility-methods.js) for buttons that are descendants of selects (something like `selectButtonHidden`) and then use that in the `isHiddenForEveryone` function.

**Comment by straker** [comment]
@ItsMeSousuke Would you have time to update the pr, or would you like us to take it over?

**Comment by ItsMeSousuke** [comment]
> @ItsMeSousuke Would you have time to update the pr, or would you like us to take it over?

I've updated this PR, please check it after that.

File: lib/commons/dom/visibility-methods.js [hunk 1]
```diff
@@ -257,6 +257,31 @@ export function detailsHidden(vNode) {
...
+  }
+
+  let node = vNode.parent;
+  while (node) {
+    if (node.props.nodeName === 'select') {
+      return true;
+    }
+
+    node = node.parent;
+  }
```
Comment (straker) [inline lib/commons/dom/visibility-methods.js]: Doing an ancestor look up on the entire ancestor tree for every button on the page is going to be a huge performance problem. Mostly because most buttons are not going to be descendants of selects so this will always navigate up to the root element for every button. 

To solve for these types of problems where we need to know if a specific ancestor exists we use `memoize` on a recursive function. That way every node on the tree only gets asked the question once (is there a select parent) and if two buttons share the same ancestor, that shared ancestor already knows the answer.

An example of this kind of ancestry lookup is [is-fixed-position](https://github.com/dequelabs/axe-core/blob/develop/lib/commons/dom/is-fixed-position.js). Lastly we make sure to use [primitive mode for memoize](https://github.com/medikoo/memoizee#primitive-mode) to speed up performance even further.

```js
export function selectButtonHidden(vNode) {
  if (vNode.props.nodeName !== 'button') {
    return false;
  }

  if (!vNode.parent) {
    return false;
  }

  return hasSelectAncestor(vNode.parent);
}

const hasSelectAncestor = memoize(function hasSelectAncestorMemoized(vNode) {
  if (vNode.props.nodeName === 'select') {
    return true;
  }

  if (!vNode.parent) {
    return false;
  }

  return hasSelectAncestor(vNode.parent);
}, { primitive: true });
```

Comment (straker) [inline lib/commons/dom/visibility-methods.js]: Instead of a loop here we can use `closest` to find the first `select` ancesotr

```suggestion
  if (closest(node, 'select')) {
    return true;
  }
```

File: test/commons/dom/is-hidden-for-everyone.js [hunk 2]
```diff
@@ -395,5 +395,81 @@ describe('dom.isHiddenForEveryone', () => {
       const actual = isHiddenForEveryone(vNode);
       assert.isFalse(actual);
     });
+
+    it('should return true for a button inside a select', () => {
```
Comment (straker) [inline test/commons/dom/is-hidden-for-everyone.js:399]: These tests should be go in their own describe block and not the `SerialVirtualNode` describe, similar to the other display tests. Can add it right after the `details` describe block.

File: lib/commons/dom/visibility-methods.js (hunk 1 above)
Comment (straker) [inline lib/commons/dom/visibility-methods.js:280]: Instead of a loop here we should use `closest`


```suggestion
  if (closest(vNode, 'select')) {
    return true;
  }
```

File: test/commons/dom/is-hidden-for-everyone.js [hunk 3]
```diff
@@ -395,5 +395,81 @@ describe('dom.isHiddenForEveryone', () => {
...
+    });
+
+    it('should return true for a select button inside an open shadow DOM', () => {
+      const vNode = queryShadowFixture(
+        '<div id="shadow"></div>',
+        '<select><button id="target"></button><option>a</option></select>'
+      );
+
+      assert.isTrue(isHiddenForEveryone(vNode));
+    });
```
Comment (straker) [inline test/commons/dom/is-hidden-for-everyone.js:473]: A shadow DOM test isn't needed unless the test in question needs to test if the content can cross shadow DOM boundaries. Select elements can't be shadow roots so we don't need to add a test for it.

**Review by straker — CHANGES_REQUESTED** [review]
Alright, so I dug into this a bit more and found I made a mistake before. I didn't realize that you needed to use the CSS style `select, ::picker(select) { appearance: base-select; }` in order to get the custom select to work. The CSS appearance changes the entire answer to when a `button` is visible or not.

When a select is not `appearance: base-select` (e.g `auto`) no button will ever show, which is what my original answer was. However when `appearance: base-select` is set then the browser will not render only the first button as the spec says, but any other button will be rendered. Also the first child element must be a button in order for it not to render, if the button is wrapped in an element then it will still render, and if the 2nd child is also a button it will still render even if the first did.

So what that means is that the `selectButtonHidden` needs to do a few things:

1) find the `select` element and check it's `appearance` CSS value (use `vNode.getComputedStylePropertyValue('appearance')`
2) if the `appearance` is not `base-select`, then any button is hidden
3) if the `appearance` is `base-select`, then only if the button is the first direct child of the select is it hidden, all others are visible. You can copy how `detailsHidden` did that check.
